src/veo_api.cpp

```diff
@@ -371,6 +457,89 @@ int veo_write_mem(veo_proc_handle *h, uint64_t dst, const void *src,
   }
 }
 
+/**
+ * @brief Copy VE/VH memory
+ *
+ * This function copies memory. When only the destination is the VE
+ * memory allocated by veo_alloc_hmem(), the data is transferred 
+ * from VH to VE. When only the source is the VE memory allocated by
+ * veo_alloc_hmem(), the data is transferred from VE to VH. When 
+ * both the source and the destination are VH memory, they are 
+ * copied on VH. When both the source and the destination are VE 
+ * memory allocated by veo_alloc_hmem(), veo_hmemcpy() returns a 
+ * failure.
+ *
+ * @param dst a pointer to destination
+ * @param src a pointer to source
+ * @param size size in byte
+ * @return zero upon success; negative upon failure.
+ */
+int veo_hmemcpy(void *dest, const void *src, size_t size)
+{
+  try
+  {
+    //std::lock_guard<std::mutex> lock(ProcHandle::__procs_mtx);
+    veo_proc_handle *h;
+    veo::ProcHandle *p;
+    int proc_ident = -1;
+    if (IS_VE(dest) && !IS_VE(src)) {
+      proc_ident = GET_PROC_IDENT(dest);
+      p = veo::ProcHandle::getProcHandle(proc_ident);
+      h = p->toCHandle();
+      return veo_write_mem(h, VIRT_ADDR_VE(dest), src, size);
+    } else if (!IS_VE(dest) && IS_VE(src)) {
+      proc_ident = GET_PROC_IDENT(src);
+      p = veo::ProcHandle::getProcHandle(proc_ident);
+      h = p->toCHandle();
+      return veo_read_mem(h, dest, VIRT_ADDR_VE(src), size);
+    } else if (!IS_VE(dest) && !IS_VE(src)) {
+      memcpy(dest, src, size);
+      return 0;
+    }
+    return -1;
+  } catch (std::out_of_range &e) {
+    VEO_ERROR("failed to get process handler: %s", e.what());
+    return -1;
+  } catch (VEOException &e) 
+  {
+    return -1;
+  }
+}
+
+/**
+ * @brief set a pointer argument
+ *
+ * This function removes the identifier from the address allocated 
+ * by veo_alloc_hmem() and passes the valid pointer to VE.
+ *
+ * @param ca veo_args
+ * @param argnum the argnum-th argument
+ * @param val a pointer to value to be set
+ * @return zero upon success; negative upon failure.
+ */
+int veo_args_set_hmem(struct veo_args *ca, int argnum, void *val)
```

Comment:
I'm sorry I didn't notice when first looking at the proposal, you caught me in a very busy time...
This call makes little sense to me. We can use anytime the `VIRT_ADDR_VE()` makro when passing an argument just normally and strip the hmem attributes. You might also add a function which does the stripping of hmem knowledge from the address. But as it is, the name of the function is misleading for me. It suggests that we pass the heterogeneous memory address, while it does the opposite.

What could make sense is passing the knowledge on hmem to the VE, such that the VE side has a chance to find out where the heterogeneous address is located (other VE/VH). This can have a value once we (maybe) add direct VE-VE transfers. Conceptionally ve-urpc could do that. But, as I commented at the beginning, then we can simply pass the hmem address as a normal argument, so no need to an API for that, either.

VEDA added device-to-device transfer initiated from the host but currently uses the host as intermediate station: device ->host->other device. There were requests on triggering VE-VE communication from inside the VE kernel, but this is a different story...